### ISSUE TITLE
[Fix bugs] Training

### ISSUE THREAD

**Issue by yjhong89** [hgt-kigam/DAS-DSVDD#1]
안녕하세요, 기택님.
다음과 같은 사항이 궁금합니다.

1. Autoencoder training loss: AutoEncoder가 훈련이 제대로 되는지 (npy 파일 말고 epoch (x축)에 따른 loss curve (y축)의 plot이면 보기 좋을 것 같습니다)
2. `c`의 범위: Latent space 상에서 `c` 값이 어느 정도인지?
3. Dataloader를 거친 데이터의 값: 이미지 하나가 모델에 들어갈 때 값의 scale이 적당하지를 보기 위함입니다. 저번에 보여주실 때 loss 값이 1로 고정되는 것으로 보아 이미지들이 거의 같은 값 (0근처)으로 들어가고 있을 확률도 있을 것 같습니다.
4. 모델 파일을 보면 1024, 224, 96, 96_100 등으로 되어 있는데 각 파일명의 이름을 알면 좋겠습니다.

**Comment by hgt-kigam** [comment]
1. AE와 SVDD 모두 이미지 플롯하는 코드를 추가했습니다.
2. 현재 Latent space의 차원이 100이며,  epoch당 100개의 c (z의 평균, z의 차원은 (input_size, 100))이 결정됨.
3. Dataloader를 통해 불러온 이미지는 0과 1사이의 분포하며, 이는 tdms(raw data format)에서 png로 변환할 때 결정됩니다.
4. 기존에 사용했던 모델로 현재는 사용하지 않습니다. 이미지의 input size에 따라 네이밍된 것이며, 1024의 경우 이미지 크기가 1024x1024일 때 사용하는 모델입니다.

1번과 2번은 훈련 진행 후 결과 첨부하겠습니다.

**Comment by hgt-kigam** [comment]
![Uploading image.png…]()
AE의 loss가 2300부근으로 수렴하는것으로 나타납니다.
제 코드에서는 AE의 parameter를 Deep-SVDD에 전파할 때 loss가 가장 작은 epoch의 parameter를 저장해서 사용합니다.
그래서 c 값을 따로 플롯해본 결과(./code/train.py의 44번 line로 생성된 c)
[-0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, -0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, -0.10000000149011612, 0.10000000149011612, -0.10000000149011612, 0.10000000149011612, 0.10000000149011612, -0.10000000149011612, -0.10000000149011612, -0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, 0.10000000149011612, -0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, -0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, 0.10000000149011612, -0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, -0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, -0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, -0.10000000149011612, 0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, -0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, -0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, -0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, -0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, 0.10000000149011612, -0.10000000149011612, -0.10000000149011612, -0.10000000149011612, 0.10000000149011612, -0.10000000149011612, 0.10000000149011612]
로 나타나며, 이러한 이유로 Deep-SVDD 모델이 1에 수렴하는 것으로 생각됩니다. 따라서 이미지에서는 AE의 loss가 정상적으로 수렴하는 것처럼 보였으나, AE가 제대로 학습되었다고 볼 수 없습니다.
제 생각으로 이런 상황이 발생된 원인으로는 학습데이터가 너무나 적다고 생각됩니다. (event image 91장, noise image 60장)
현재 190716~190723 1주일치의 자료 중 인접한 지진계 (학계리[HAK] 지진계)에서 진폭 변화가 육안으로 관찰되는 시점의 자료를 사용하여 매우 적은 수로 학습에 사용중인데 다른 시간대의 DAS 자료와 이미지 증대 기법을 사용하고, 공유해주신 SAD 코드를 활용(train과 test의 nor, ano split ratio 차이)하여 추가적으로 학습 진행해보겠습니다.

**Comment by hgt-kigam** [comment]
![Uploading image.png…]()
.

**Comment by yjhong89** [comment]
- 먼저 auto encoder가 수렴은 해도 loss 값 자체가 너무 커보입니다. 경험 상 loss가 몇 백대로 더 내려가야 수렴하는 듯 합니다.
- 여기 코드 https://github.com/hgt-kigam/DAS-DSVDD/blob/f1f2e90866832eb22d027f23d5bf6454a8753fd2/code/train.py#L75 를 보면 `c` 값이 epsilon 보다 작아서 강제로 0.1로 setting 되는 것처럼 보이네요. 
- 제가 잘 이해하지 못한 것 같은데, 기존과 계속 이미지 사이즈는 동일한 것 아닌가요? (convolution filter size만 변한 게 아닌가요?)

**Comment by hgt-kigam** [comment]
c가 강제로 수렴되는 것 같다는 대표님 의견에 동의합니다. 

9월 발표했던 모델과 비교하면 convolution filter size, input image size, train data size 전부 달라졌습니다.
기존 모델은 noise 이미지가 훨씬 많았습니다. (대략 event 2,000장, noise 600,000장)

이러한 이유로 모델의 complexity가 학습데이터셋보다 높아서 일종의 과적합으로 생각하는 것이 옳을까요?

decoder 마지막 CNN 블록의 ReLU + BN를 제거하여 훈련해보겠습니다.

코멘트 감사합니다!

**Comment by yjhong89** [comment]
- 넵, 제 생각에는 전형적인 over-fitting 현상 같습니다. 
  - augmentation을 적용해야 할 것 같습니다.
- 넵 그렇게 훈련해보고 진행사항 보면 좋을 것 같습니다. (안 쓰는 코드도 정리하면 좋을 것 같습니다)

**Comment by hgt-kigam** [comment]
<img width="648" alt="image" src="https://github.com/hgt-kigam/DAS-DSVDD/assets/140587931/a80af801-9822-440f-83ac-faf5b2169103">
상단 이미지: AE (Loss: 58.665071)
하단 이미지: SVDD (Loss: 0.819667)

<img width="443" alt="image" src="https://github.com/hgt-kigam/DAS-DSVDD/assets/140587931/ac94ceb2-822b-48db-865c-78f287e10ab9">
normal class는 0으로 지정하고 학습했습니다.

**Comment by yjhong89** [comment]
음.. autoencoer가 너무 복잡한 것 같으니 batch normalization 을 빼고 간단한 모델로 테스트 해보시는 것은 어떠신가요?
- https://github.com/SuchismitaSahu1993/Autoencoder-on-MNIST-in-Pytorch/blob/master/Autoencoder.py
